src/explore-education-statistics-admin/src/contexts/AuthContext.tsx

```diff
@@ -43,80 +41,343 @@ interface Props {
+  invites. At this point they will be redirected to either the "No Invites"
+  page or the "Expired Invites" page, and this flag will be set to true so
+  that they can successfully see the page. */
+  readyToRenderChildren: boolean;
+
+  /**
+   * The user profile and their permissions, if a user has successfully
+   * logged in or registered using the local login / registration endpoint.
+   */
   user?: User;
+
+  /**
+   * The post-login "state" object returned from the Identity Provider
+   * post-login redirect. If the user has loaded the SPA via a redirect from the
+   * Identity Provider, this "state" object will hold a "returnUrl" that we
+   * requested be provided when the user originally clicked on the "Sign in"
+   * button and was sent to the Identity Provider to log in.
+   */
+  loginRedirectState?: PostLoginState;
+
+  /**
+   * A path to redirect the user to prior to rendering the children of this
+   * component, if any redirect is required.
+   */
+  redirect?: string;
 }
 
 export const AuthContextProvider = ({ children }: Props) => {
-  const [state, setState] = useState<State>({ ready: false });
+  const [state, setState] = useState<State>({
+    readyToRenderChildren: false,
+  });
+
+  const {
+    instance: msalInstance,
+    accounts,
+    inProgress: authenticationInProgress,
+  } = useMsal();
+
+  const loginInProgress = authenticationInProgress !== 'none';
+
+  const history = useHistory();
+  const location = useLocation();
+
+  // Register a listener for the post-login callback when returning from
+  // the external Identity Provider login. This will allow us to access the
+  // "state" object from the post-login response and extract the returnUrl (if
+  // any) that we set during the loginRedirect() call in msal.ts.
+  useEffect(() => {
+    msalInstance.handleRedirectPromise().then(loginResult => {
+      if (loginResult) {
+        const loginRedirectState = loginResult.state
+          ? (JSON.parse(loginResult.state) as PostLoginState)
+          : undefined;
+        if (loginRedirectState) {
+          setState(previousState => ({
+            ...previousState,
+            loginRedirectState,
+          }));
+        }
+      }
+    });
+    // We want to register this callback handler once and only once.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  // Check to see if we have an explicit request to redirect to a particular
+  // route. This can occur if a login failure occurs or if login is successful
+  // and an explicit redirect URL has been specified post-login.
+  useEffect(() => {
+    if (state.redirect) {
+      logger.info(
+        `AuthContext: redirect to ${state.redirect} requested. Setting user ` +
+          'ready to use service.',
+      );
+      setState(previousState => ({
+        ...previousState,
+        readyToRenderChildren: true,
+      }));
+      logger.info(`AuthContext: redirecting to ${state.redirect}.`);
```

Comment:
There's a lot of logging in this file, do we really need it all?

Reply:
Good point - I'd quite like to keep it til the last minute in case bugs are found during the testing of this, but I'll put it behind a flag

Reply:
I've stuck a "verboseLogging" flag that defaults to false.  If anyone needs to debug it, we can set it to true in App.tsx.  The check and the logging are set up as a "log" function in AuthContext. Does it need to be memoized?  Ta.

Reply:
Cool. I don't think it needs to be memoised. You could change `verboseLogging` to be optional and default to `false` then we wouldn't need to set it in `App.tsx` (unless you want to turn logging on locally).

Reply:
Cool thanks, I have made `verboseLogging` optional as you suggested.

I also converted the "log" function to a `useCallback` and included it in the main dependency array.  Originally I had excluded it from the dependency array because it was causing endless re-renders, but this way everything should be satisfied.